clientconn_test.go

```diff
@@ -1079,6 +1080,113 @@ func (s) TestDefaultServiceConfig(t *testing.T) {
 	testDefaultServiceConfigWhenResolverReturnInvalidServiceConfig(t, r, addr, js)
 }
 
+type testSubscriber struct {
+	onMsgCh chan connectivity.State
+}
+
+func newTestSubscriber() *testSubscriber {
+	return &testSubscriber{onMsgCh: make(chan connectivity.State, 1)}
+}
+
+func (ts *testSubscriber) OnMessage(msg interface{}) {
+	select {
+	case ts.onMsgCh <- msg.(connectivity.State):
+	default:
+	}
+}
+
+func (s) TestReportStateChangesToSubscriber(t *testing.T) {
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening. Err: %v", err)
+	}
+	defer lis.Close()
+
+	done := make(chan struct{})
+	sent := make(chan struct{})
+	dialDone := make(chan struct{})
+
+	// Launch the server.
+	go func() {
+		defer func() {
+			close(done)
+		}()
+		conn, err := lis.Accept()
+		if err != nil {
+			t.Errorf("Error while accepting. Err: %v", err)
+			return
+		}
+		defer conn.Close()
+		// Sleep for a little bit to make sure that Dial on client
+		// side blocks until settings are received.
+		time.Sleep(100 * time.Millisecond)
+		framer := http2.NewFramer(conn, conn)
+		close(sent)
+		if err := framer.WriteSettings(http2.Setting{}); err != nil {
+			t.Errorf("Error while writing settings. Err: %v", err)
+			return
+		}
+		<-dialDone // Close conn only after dial returns.
+	}()
```

Comment:
Please write this as an end-to-end style test in the `grpc/test` package. 
- Create a new file `connectivity_state_updates_test.go`
- Create a test backend service. See: https://github.com/grpc/grpc-go/blob/master/internal/stubserver/stubserver.go#L219
- Create a manual resolver
- Create a ClientConn that dials the above test service and uses the manual resolver
- Subscribe to connectivity state updates
- Ensure the initial state is as expected
- Push the address of the test backend through the manual resolver
- Ensure that you can make an RPC to the test backend
- Ensure that all expected states upto READY are received by the subscriber
- Close the clientConn
- Ensure that SHUTDOWN is received by the subscriber.


See https://github.com/grpc/grpc-go/blob/master/test/idleness_test.go to get an idea of how to create the test backend, how to create the manual resolver, how to push updates using the manual resolver etc.

Please let me know if you have any questions here. Thanks.